src/lfortran/parser/semantics.h

```diff
@@ -63,6 +63,23 @@ static inline char* name2char(const ast_t *n)
     return down_cast2<Name_t>(n)->m_id;
 }
 
+static inline void check_subroutine_names(char* n1, char* n2, Location &loc) {
+    if (strcmp(n1, n2) != 0) {
+        throw LCompilers::LFortran::parser_local::ParserError(
+            "End subroutine name does not match subroutine name", loc);
+    }
+}
+
+static inline char* name2char_check(const ast_t *n1, const ast_t *n2,
+        Location &loc) {
+    char* n1c = down_cast2<Name_t>(n1)->m_id;
+    if(n2) {
+        char* n2c = down_cast2<Name_t>(n2)->m_id;
+        check_subroutine_names(n1c, n2c, loc);
+    }
```

Comment:
```suggestion
    char* n1c = name2char(n1);
    if(n2) {
        char* n2c = name2char(n2);
        check_subroutine_names(n1c, n2c, loc);
    }
```
You can do this check in the subsequent PRs.